File: 213.打家劫舍-ii.py
#
# @lc app=leetcode.cn id=213 lang=python3
#
# [213] 打家劫舍 II
#

# @lc code=start
class Solution:
    def rob(self, nums: List[int]) -> int:
        if len(nums) == 1:
            return nums[0]
        unsteal = [0, nums[1]]
        steal = [nums[0], nums[0]]
        for i in range(2, len(nums)):
            if i != len(nums)-1:
                unsteal = self.getMax(unsteal, nums[i])
                steal = self.getMax(steal, nums[i])
            else:
                unsteal = self.getMax(unsteal, nums[i])
        return max(max(unsteal), max(steal))
    
    def getMax(self, steal: List[int], num:int):
        tmp = max(steal[0]+num, steal[1])
        steal[0] = steal[1]
        steal[1] = tmp
        return steal

# @lc code=end

File: 213.打家劫舍-ii.java
import java.util.Arrays;

/*
 * @lc app=leetcode.cn id=213 lang=java
 *
 * [213] 打家劫舍 II
 */

// @lc code=start
class Solution {
    public int rob(int[] nums) {
        if(nums.length==1){
            return nums[0];
        }
        if(nums.length==2){
            return Math.max(nums[0], nums[1]);
        }
        return Math.max(robNums(Arrays.copyOfRange(nums, 0, nums.length-1)), robNums(Arrays.copyOfRange(nums, 1, nums.length)));
    }
    public int robNums(int[] nums){
        int[] dp = new int[2];
        dp[0] = nums[0];
        dp[1] = Math.max(nums[0], nums[1]);
        for(int i=2; i<nums.length; i++){
            int tmp = Math.max(dp[0]+nums[i], dp[1]);
            dp[0] = dp[1];
            dp[1] = tmp;
        }
        return Math.max(dp[0], dp[1]);
    }
}
// @lc code=end

